contracts/atomic-swap/src/msg.rs

```diff
@@ -0,0 +1,84 @@
+use cosmwasm_std::{Coin, HumanAddr};
+use schemars::JsonSchema;
+use serde::{Deserialize, Serialize};
+
+#[derive(Serialize, Deserialize, JsonSchema)]
+pub struct InitMsg {}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum HandleMsg {
+    Create(CreateMsg),
+    /// Release sends all tokens to the recipient.
+    Release {
+        id: String,
+        /// This is the preimage, must be exactly 32 bytes in hex (64 chars)
+        /// to release: sha256(from_hex(preimage)) == from_hex(hash)
+        preimage: String,
+    },
+    /// Refund returns all remaining tokens to the original sender,
+    Refund {
+        id: String,
+    },
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+pub struct CreateMsg {
+    /// id is a human-readable name for the swap to use later.
+    /// 3-20 bytes of utf-8 text
+    pub id: String,
+    /// This is hex-encoded sha-256 hash of the preimage (must be 32*2 = 64 chars)
+    pub hash: String,
+    /// If approved, funds go to the recipient
+    pub recipient: HumanAddr,
+    /// You can set a last time or block height the contract is valid at.
+    /// If *either* is non-zero and below current state, the contract is considered expired,
+    /// and will be returned to the original funder.
+    pub end_height: u64,
+    pub end_time: u64,
+}
+
+pub fn is_valid_name(name: &str) -> bool {
+    let bytes = name.as_bytes();
+    if bytes.len() < 3 || bytes.len() > 20 {
+        return false;
+    }
+    true
+}
+
+#[derive(Serialize, Deserialize, Clone, Debug, PartialEq, JsonSchema)]
+#[serde(rename_all = "snake_case")]
+pub enum QueryMsg {
+    /// Show all open swaps. Return type is ListResponse.
+    List {},
+    /// Returns the details of the named swap, error if not created.
+    /// Return type: DetailsResponse.
+    Details { id: String },
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct ListResponse {
+    /// List all open swap ids
+    pub swaps: Vec<String>,
+}
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, JsonSchema, Debug)]
+pub struct DetailsResponse {
+    /// Id of this swap
+    pub id: String,
+    /// This is hex-encoded sha-256 hash of the preimage (must be 32*2 = 64 chars)
+    pub hash: String,
+    /// If released, funds go to the recipient
+    pub recipient: HumanAddr,
+    /// If refunded, funds go to the source
+    pub source: HumanAddr,
+    /// When end height greater than zero and block height exceeds this value, the swap is expired.
+    /// Once a swap is expired, it can be returned to the original source (via "refund").
+    pub end_height: u64,
```

Comment:
Again, if we use `cw0::Expiration` in the internals, we can use it in this API as well